### ISSUE TITLE
ceph: Dashboard throws errors

### ISSUE THREAD

**Issue by mmack** [rook/rook#3106]
**Is this a bug report or feature request?**
* Bug Report

**Deviation from expected behavior:**
Dashboard home does not work

**Expected behavior:**
Dashboard home page should work

**How to reproduce it (minimal and precise):**
Update ceph from v14.2.0-20190410 to v14.2.1-20190430 

**Log on the mgr pod:**
```
::ffff:100.192.13.24 - - [03/May/2019:11:10:03] "GET /api/health/minimal HTTP/1.1" 500 1646 "https://rook-ceph-dev-primary.k8s.conmpany.com/" "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Ubuntu Chromium/73.0.3683.86 Chrome/73.0.3683.86 Safari/537.36"
debug 2019-05-03 11:10:03.630 7f0474a6e700  0 log_channel(cluster) log [DBG] : pgmap v193: 200 pgs: 100 active+undersized, 100 active+clean; 0 B data, 278 MiB used, 3.8 TiB / 3.9 TiB avail
debug 2019-05-03 11:10:05.631 7f0474a6e700  0 log_channel(cluster) log [DBG] : pgmap v194: 200 pgs: 100 active+undersized, 100 active+clean; 0 B data, 278 MiB used, 3.8 TiB / 3.9 TiB avail
debug 2019-05-03 11:10:07.631 7f0474a6e700  0 log_channel(cluster) log [DBG] : pgmap v195: 200 pgs: 100 active+undersized, 100 active+clean; 0 B data, 278 MiB used, 3.8 TiB / 3.9 TiB avail
::ffff:100.192.13.24 - - [03/May/2019:11:10:07] "GET /api/summary HTTP/1.1" 200 257 "https://rook-ceph-dev-primary.k8s.conmpany.com/" "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Ubuntu Chromium/73.0.3683.86 Chrome/73.0.3683.86 Safari/537.36"
::ffff:100.192.13.24 - - [03/May/2019:11:10:07] "GET /api/feature_toggles HTTP/1.1" 200 66 "https://rook-ceph-dev-primary.k8s.conmpany.com/" "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Ubuntu Chromium/73.0.3683.86 Chrome/73.0.3683.86 Safari/537.36"
::ffff:100.192.13.24 - - [03/May/2019:11:10:07] "GET /api/prometheus/notifications?from=last HTTP/1.1" 200 22 "https://rook-ceph-dev-primary.k8s.conmpany.com/" "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Ubuntu Chromium/73.0.3683.86 Chrome/73.0.3683.86 Safari/537.36"
debug 2019-05-03 11:10:08.072 7f0438724700  0 mgr[rook] Completion <read op> threw an exception:
Traceback (most recent call last):
  File "/usr/share/ceph/mgr/rook/module.py", line 191, in wait
    c.execute()
  File "/usr/share/ceph/mgr/rook/module.py", line 57, in execute
    self._result = self.cb()
  File "/usr/share/ceph/mgr/rook/module.py", line 132, in <lambda>
    return RookReadCompletion(lambda: f(*args, **kwargs))
  File "/usr/share/ceph/mgr/rook/module.py", line 348, in describe_service
    assert service_type in ("mds", "osd", "mgr", "mon", "nfs", None), service_type + " unsupported"
AssertionError: iscsi unsupported
[03/May/2019:11:10:08] HTTP 
Request Headers:
  AUTHORIZATION: Bearer XXXXXX
  REFERER: https://rook-ceph-dev-primary.k8s.conmpany.com/
  X-FORWARDED-HOST: rook-ceph-dev-primary.k8s.conmpany.com
  Remote-Addr: ::ffff:100.192.13.24
  X-REAL-IP: 127.0.0.1
  USER-AGENT: Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Ubuntu Chromium/73.0.3683.86 Chrome/73.0.3683.86 Safari/537.36
  CONNECTION: close
  COOKIE: _ga=GA1.2.1677006237.1530686008; session_id=d057bedfc9b0909a5da6956eb3d4b8650dce540e
  ACCEPT: application/json, text/plain, */*
  PRAGMA: no-cache
  X-FORWARDED-PROTO: https
  X-ORIGINAL-URI: /api/health/minimal
  HOST: rook-ceph-dev-primary.k8s.conmpany.com
  CACHE-CONTROL: no-cache
  X-SCHEME: https
  ACCEPT-LANGUAGE: en-US,en;q=0.9,de;q=0.8
  X-FORWARDED-FOR: 127.0.0.1
  X-FORWARDED-PORT: 443
  X-REQUEST-ID: b8ebc25bb44f0b89f890092e3b508d89
  ACCEPT-ENCODING: gzip, deflate, br
[03/May/2019:11:10:08] HTTP Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/cherrypy/_cprequest.py", line 656, in respond
    response.body = self.handler()
  File "/usr/lib/python2.7/site-packages/cherrypy/lib/encoding.py", line 188, in __call__
    self.body = self.oldhandler(*args, **kwargs)
  File "/usr/lib/python2.7/site-packages/cherrypy/_cptools.py", line 221, in wrap
    return self.newhandler(innerfunc, *args, **kwargs)
  File "/usr/share/ceph/mgr/dashboard/services/exception.py", line 88, in dashboard_exception_handler
    return handler(*args, **kwargs)
  File "/usr/lib/python2.7/site-packages/cherrypy/_cpdispatch.py", line 34, in __call__
    return self.callable(*self.args, **self.kwargs)
  File "/usr/share/ceph/mgr/dashboard/controllers/__init__.py", line 649, in inner
    ret = func(*args, **kwargs)
  File "/usr/share/ceph/mgr/dashboard/controllers/health.py", line 197, in minimal
    return self.health_minimal.all_health()
  File "/usr/share/ceph/mgr/dashboard/controllers/health.py", line 62, in all_health
    result['iscsi_daemons'] = self.iscsi_daemons()
  File "/usr/share/ceph/mgr/dashboard/controllers/health.py", line 126, in iscsi_daemons
    gateways = IscsiGatewaysConfig.get_gateways_config()['gateways']
  File "/usr/share/ceph/mgr/dashboard/services/iscsi_config.py", line 93, in get_gateways_config
    for instance in instances:
TypeError: 'NoneType' object is not iterable

debug 2019-05-03 11:10:08.073 7f0438724700  0 mgr[dashboard] [03/May/2019:11:10:08] HTTP Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/cherrypy/_cprequest.py", line 656, in respond
    response.body = self.handler()
  File "/usr/lib/python2.7/site-packages/cherrypy/lib/encoding.py", line 188, in __call__
    self.body = self.oldhandler(*args, **kwargs)
  File "/usr/lib/python2.7/site-packages/cherrypy/_cptools.py", line 221, in wrap
    return self.newhandler(innerfunc, *args, **kwargs)
  File "/usr/share/ceph/mgr/dashboard/services/exception.py", line 88, in dashboard_exception_handler
    return handler(*args, **kwargs)
  File "/usr/lib/python2.7/site-packages/cherrypy/_cpdispatch.py", line 34, in __call__
    return self.callable(*self.args, **self.kwargs)
  File "/usr/share/ceph/mgr/dashboard/controllers/__init__.py", line 649, in inner
    ret = func(*args, **kwargs)
  File "/usr/share/ceph/mgr/dashboard/controllers/health.py", line 197, in minimal
    return self.health_minimal.all_health()
  File "/usr/share/ceph/mgr/dashboard/controllers/health.py", line 62, in all_health
    result['iscsi_daemons'] = self.iscsi_daemons()
  File "/usr/share/ceph/mgr/dashboard/controllers/health.py", line 126, in iscsi_daemons
    gateways = IscsiGatewaysConfig.get_gateways_config()['gateways']
  File "/usr/share/ceph/mgr/dashboard/services/iscsi_config.py", line 93, in get_gateways_config
    for instance in instances:
TypeError: 'NoneType' object is not iterable
```

**ceph's status (i went to the dashboard to actually increase the pgs):**
```
  cluster:
    id:     a07354be-df7f-43d4-b292-733187b45ee3
    health: HEALTH_WARN
            Degraded data redundancy: 100 pgs undersized
            too few PGs per OSD (25 < min 30)
 
  services:
    mon:        5 daemons, quorum a,b,c,d,e (age 12m)
    mgr:        a(active, since 11m)
    osd:        32 osds: 32 up (since 8m), 32 in (since 8m)
    rbd-mirror: 3 daemons active (15161, 35273, 44884)
 
  data:
    pools:   2 pools, 200 pgs
    objects: 0 objects, 0 B
    usage:   32 GiB used, 3.8 TiB / 3.9 TiB avail
    pgs:     100 active+clean
             100 active+undersized
```

**Environment**:
* OS (e.g. from /etc/os-release):
* Kernel (e.g. `uname -a`):
* Cloud provider or hardware configuration:
* Rook version (use `rook version` inside of a Rook Pod):
* Kubernetes version (use `kubectl version`):
* Kubernetes cluster type (e.g. Tectonic, GKE, OpenShift):
* Storage backend status (e.g. for Ceph use `ceph health` in the [Rook Ceph toolbox](https://rook.io/docs/rook/master/ceph-toolbox.html)):

**Comment by monester** [comment]
Have the same problem with dashboard module. Rook version is 1.0.0.

**Comment by marcstreeter** [comment]
I'm running a fresh install of v1.0 having followed the flex volume steps for RKE and I notice this error in the browser console
```
{
    "status": "500 Internal Server Error",
    "version": "3.2.2",
    "detail": "The server encountered an unexpected condition which prevented it from fulfilling the request.",
    "traceback": "Traceback (most recent call last):\n  File \"/usr/lib/python2.7/site-packages/cherrypy/_cprequest.py\", line 656, in respond\n    response.body = self.handler()\n  File \"/usr/lib/python2.7/site-packages/cherrypy/lib/encoding.py\", line 188, in __call__\n    self.body = self.oldhandler(*args, **kwargs)\n  File \"/usr/lib/python2.7/site-packages/cherrypy/_cptools.py\", line 221, in wrap\n    return self.newhandler(innerfunc, *args, **kwargs)\n  File \"/usr/share/ceph/mgr/dashboard/services/exception.py\", line 88, in dashboard_exception_handler\n    return handler(*args, **kwargs)\n  File \"/usr/lib/python2.7/site-packages/cherrypy/_cpdispatch.py\", line 34, in __call__\n    return self.callable(*self.args, **self.kwargs)\n  File \"/usr/share/ceph/mgr/dashboard/controllers/__init__.py\", line 649, in inner\n    ret = func(*args, **kwargs)\n  File \"/usr/share/ceph/mgr/dashboard/controllers/health.py\", line 197, in minimal\n    return self.health_minimal.all_health()\n  File \"/usr/share/ceph/mgr/dashboard/controllers/health.py\", line 62, in all_health\n    result['iscsi_daemons'] = self.iscsi_daemons()\n  File \"/usr/share/ceph/mgr/dashboard/controllers/health.py\", line 126, in iscsi_daemons\n    gateways = IscsiGatewaysConfig.get_gateways_config()['gateways']\n  File \"/usr/share/ceph/mgr/dashboard/services/iscsi_config.py\", line 93, in get_gateways_config\n    for instance in instances:\nTypeError: 'NoneType' object is not iterable\n"
}
```

**Comment by vitobotta** [comment]
Same here after upgrading to v1.0 - `AssertionError: iscsi unsupported`

**Comment by noahdesu** [comment]
confirmed that this is an issue in my environment.

**Comment by noahdesu** [comment]
@ricardoasmarques @jtlayton @sebastian-philipp this is related to dashboard and orchestrators. I've filed an issue upstream in ceph, too: http://tracker.ceph.com/issues/39586

The issue here is that orchestrator manager modules (rook and deepsea) are asserting that iscsi is not a type of service they deal with, but the dashboard appears to be expecting a result (perhaps an empty result, but not an assertion failure :)

```
  File "/usr/share/ceph/mgr/rook/module.py", line 348, in describe_service
    assert service_type in ("mds", "osd", "mgr", "mon", "nfs", None), service_type + " unsupported"
AssertionError: iscsi unsupported
```

`get_gateways_config` invokes the remote rook module method above resulting in trying to iterate over NoneType.

```
  File "/usr/share/ceph/mgr/dashboard/controllers/health.py", line 126, in iscsi_daemons
    gateways = IscsiGatewaysConfig.get_gateways_config()['gateways']
  File "/usr/share/ceph/mgr/dashboard/services/iscsi_config.py", line 93, in get_gateways_config
    for instance in instances:
TypeError: 'NoneType' object is not iterable
```

It would seem that either the orchestrator modules should return something like an empty list.

**Comment by noahdesu** [comment]
the only workaround i've managed to come up with is to create a new copy of the built-in administrator role, remove iscsi permissions from this role, and assign this new role to the admin. presumably once the issue is fixed upstream one can delete the new role and reassign the administrator role to the admin user.

```bash
ceph dashboard ac-role-create admin-no-iscsi

for scope in dashboard-settings log rgw prometheus grafana nfs-ganesha manager hosts rbd-image config-opt rbd-mirroring cephfs user osd pool monitor; do
    ceph dashboard ac-role-add-scope-perms admin-no-iscsi ${scope} create delete read update;
done

ceph dashboard ac-user-set-roles admin admin-no-iscsi
```

**Comment by vitobotta** [comment]
Thanks @noahdesu , the workaround works for me.

**Comment by Keithsc** [comment]
Workaround works for me also, thanks @noahdesu

**Comment by sebastian-philipp** [comment]
another idea would be to instead run 
```
ceph orchestrator set backend ""
```
which disables the rook integration into the Ceph Dashboard.

**Comment by NickUfer** [comment]
> the only workaround i've managed to come up with is to create a new copy of the built-in administrator role, remove iscsi permissions from this role, and assign this new role to the admin. presumably once the issue is fixed upstream one can delete the new role and reassign the administrator role to the admin user.
> 
> ```shell
> ceph dashboard ac-role-create admin-no-iscsi
> 
> for scope in dashboard-settings log rgw prometheus grafana nfs-ganesha manager hosts rbd-image config-opt rbd-mirroring cephfs user osd pool monitor; do
>     ceph dashboard ac-role-add-scope-perms admin-no-iscsi ${scope} create delete read update;
> done
> 
> ceph dashboard ac-user-set-roles admin admin-no-iscsi
> ```

Thanks that worked. :+1: What was the default role called in case I want to remove this workaround once it is fixed?

**Comment by noahdesu** [comment]
@NickUfer `administrator`

**Comment by sebastian-philipp** [comment]
I'm going to fix this in the dashboard.

**Comment by dimm0** [comment]
Are we still waiting for the fix? Better not to update in rook 1.0.1?

**Comment by noahdesu** [comment]
If you are running 1.0.0 already, I don't think there is any reason to avoid 1.0.1. The fix for this still has to be merged upstream in Ceph and have containers built etc... though there are some workarounds listed above in this issue. @travisn  ?

**Comment by dimm0** [comment]
I'm asking about the update to nautilus
Not clear if it's easy to return to original state after those workarounds

**Comment by noahdesu** [comment]
I don't have any reason to think that these workarounds would be difficult to undo.

**Comment by noahdesu** [comment]
They'll involve some commands in the toolbox, and maybe a mgr pod restart.

**Comment by chatman** [comment]
I just installed Kubernetes, followed https://rook.io/docs/rook/v1.0/ceph-quickstart.html and ran into this same (500 Internal Server Error) issue with the Dashboard.
1) When will this be fixed?
2) How do I execute the workaround mentioned here?
Thanks!

**Comment by noahdesu** [comment]
> On May 27, 2019, at 10:31, Ishan Chattopadhyaya <notifications@github.com> wrote:
> 
> I just installed Kubernetes, followed https://rook.io/docs/rook/v1.0/ceph-quickstart.html and ran into this same (500 Internal Server Error) issue with the Dashboard.
> 
> When will this be fixed?
The fix for this I believe is merged upstream so the fix should be in the next point release probably max month or two.
> How do I execute the workaround mentioned here?
Thebcommands can be run in the toolbox container. Instructions for that are in the rook docs on the ceph toolbox page.

**Comment by noahdesu** [comment]
https://tracker.ceph.com/issues/39586

**Comment by sebastian-philipp** [comment]
https://github.com/ceph/ceph/pull/26341 still needs backporting to Nautilus, which is blocked by https://github.com/ceph/ceph/pull/28040

**Comment by muhlba91** [comment]
I am running into the same issue with `v1.0.2` and ceph version `v14.2.1-20190430`. Any information when we can expect a fix to be released and how to update? :)

**Comment by Amos-85** [comment]
same for me in version v14.2.1-20190430

**Comment by noahdesu** [comment]
These fixes seem to have been merged into upstream Ceph. It takes a while before they are backported, point releases cut, and new containers are built.

**Comment by sebastian-philipp** [comment]
https://github.com/ceph/ceph/pull/28552

**Comment by dimm0** [comment]
Are we still waiting here?

**Comment by nabokihms** [comment]
PR was merged 17 days ago, so... I think now We are waiting for new docker nautilis image here https://hub.docker.com/r/ceph/ceph/tags

**Comment by muhlba91** [comment]
I can confirm it works with the new nautilus image `v14.2.2-20190722`.

**Comment by megian** [comment]
Can confirm it too. Dashboard is working with the image `v14.2.2-20190722`, without any configuration changes.

**Comment by mumrau** [comment]
It would maybe interesting to give the commands for the `rollback` of @noahdesu workaround which has probably been used my tons of people.

**Comment by leojonathanoh** [comment]
> 
> 
> Can confirm it too. Dashboard is working with the image `v14.2.2-20190722`, without any configuration changes.

Yes, dashboard no longer throws errors in `v14.2.2-20190722`. So no longer need this workaround if you're using that image

**Comment by noahdesu** [comment]
@mumrau thanks! yes, that looks like the correct way to revert the workaround.

CC @jmolmo

**Comment by rajha-korithrien** [comment]
When will this make it into a patch release?
I just updated to Rook ```v1.0.6``` from a previous ```v1.0.1``` deployment but I find I am still using ceph image ```ceph/ceph:v14.2.1-20190430```.

Have I misunderstood and I need to take explicit action to cause my ceph deployment to use ```v14.2.2-20190722```?

**Comment by King6moh** [comment]
@rajha-korithrien there is a separation between the Operator and the Ceph versioning as of Rook 0.9.0+. See https://rook.io/docs/rook/v1.0/ceph-upgrade.html specifically the last section that talks about Ceph Version Upgrades. This was introduced back in 0.9.0 and refined over the next few releases. Here's the blog post about it: https://blog.rook.io/what-version-of-ceph-would-you-like-with-that-8275f487cdae

**Comment by mumrau** [comment]
@rajha-korithrien in short words, yes you should migrate your CephCluster to cephVersion `v14.2.2-20190722`.

**Comment by rajha-korithrien** [comment]
@King6moh @mumrau It is clear now thank you for taking the time to help!
I can confirm that after following the provided instructions for updating the Ceph image version my dashboard works as expected.
